virtdbg/vmx/vmm.c

```diff
@@ -0,0 +1,282 @@
+#include <vmx/vmm.h>
+#include <mm/pmm.h>
+#include <arch/intrin.h>
+#include <stdint.h>
+#include <vmx/ept.h>
+#include <stddef.h>
+#include <util/except.h>
+#include <virtdbg.h>
+
+//enable vmx operation
+err_t vmxon() {
+    err_t err = NO_ERROR;
+    uint32_t* vmxon_region = pallocz_aligned(0x1000, 0x1000);
+
+    size_t control = __rdmsr(0x3a);
+    if((control & (0x1 | 0x4)) != (0x1 | 0x04)) {
+        //Enabled outside of SMX and lock bit.
+        __wrmsr(0x3a, control | 0x1 | 0x4);
+    }
+
+    uint64_t cr0;
+    asm volatile ("mov %%cr0, %0" : "=r" (cr0));
+    cr0 &= __rdmsr(0x487);
+    cr0 |= __rdmsr(0x486);
+    asm volatile ("mov %0, %%cr0" : : "r" (cr0));
+
+    uint64_t cr4;
+    asm volatile ("mov %%cr4, %0" : "=r" (cr4));
+    cr4 |= 1 << 13;
+    cr4 &= __rdmsr(0x489);
+    cr4 |= __rdmsr(0x488);
+    asm volatile ("mov %0, %%cr4" : : "r" (cr4));
+
+    //Set vmx revision
+    uint32_t vmx_revision = __rdmsr(0x480);
+    *vmxon_region = vmx_revision;
+    uint16_t successful = 0;
+    asm volatile(
+        "vmxon %1;"
+        "movq $0, %%rdx;"
+        "success:"
+        "movq $1, %%rdx;"
+        :"=d"(successful)
+        :"m"(vmxon_region)
+        :"memory", "cc"
+    );
+
+    if (!successful) {
+        err = ERROR_UNSUPPORTED;
+    }
+
+    return err;
+}
+
+static inline int vmptrld(uintptr_t vmcs) {
```

Comment:
In all of these, what is the return value? If it's an error then have it return err_t and use CHECK